### ISSUE TITLE
provisiond: decommission of network fails if container is attached to it

### ISSUE THREAD

**Issue by zaibon** [threefoldtech/zos#311]
```
[-] provisiond: 11:55AM INF reservation expired created=2019-10-07T12:55:36+02:00 duration=1h0m0s id=r-0 type=network[  617.211771] b-5CFQvzwsmVtGM: port 2(veth38ca695f) entered disabled state

[-] provisiond: 11:55AM INF start decommissioning reservation duration=1h0m0s id=r-0 type=network
[-] capacityd: {"level":"info","time":"2019-10-07T11:55:40Z","message":"send heart-beat to BCDB"}
[  617.221010] device vethbfb0712b left promiscuous mode
[  617.222233] b-5CFQvzwsmVtGM: port 1(vethbfb0712b) entered disabled state
[-] networkd: 11:55AM ERR failed to delete network resource namespace error="failed to unmount NS: at /var/run/netns/n-5CFQvzwsmVtGM: device or resource busy" namespace=n-5CFQvzwsmVtGM
[-] provisiond: 11:55AM INF reservation decommission successful duration=1h0m0s id=r-0 type=network
```

**Comment by zaibon** [comment]
This is a tricky one cause there is a relation between a container and the network it uses.
I see 2 cases:

- a network expires while some container are still using it: This will lead to left over on the system cause the network namespace and interfaces for the network resource cannot be properly deleted.

- some container are running and attached to a network that has been decommissioned. Then the same network is re-provisioned again: Here the container won't get connectivity back when the network is re-created.

So my idea for this is to always check when trying to decommission a network that there is no container attached to it. If it is the case. Then the networkd won't delete the network resource but just update the nft table to stop all connectivity.
So in that case there are 2 possible outcome of this.
1. At some point all container using the network expired and then we can delete the network resource
2. The network is re-created, this would trigger the update of the nft to restart connectivity.

Practically here is what would need to change in the code:
- Add a method on the [network_resource type](https://github.com/threefoldtech/zos/blob/2ad0b43e047a72755ee58c1a5b9969cd8bc78cc2/pkg/network/nr/net_resource.go#L29-L34) that check if any container is connected to it. We could implement that by checking if there is any veth pair attach to the bridge.
- Update the [Networker.DeleteNR](https://github.com/threefoldtech/zos/blob/2ad0b43e047a72755ee58c1a5b9969cd8bc78cc2/pkg/network/networker.go#L407) method to check if there is any container attach to it and base on the or delete the NR or update the nft table.
- Update the [Networker.CreateNR](https://github.com/threefoldtech/zos/blob/2ad0b43e047a72755ee58c1a5b9969cd8bc78cc2/pkg/network/networker.go#L407) to handle the case where the NR already exists and we just need to update the nft table.

**Comment by delandtj** [comment]
well when a networks expires , the NRs associated to them expire all at the same time. So yes, a network that __has__ expired should first see if there are sill workloads associated to them. 
I think it would be better to never expire the network as it is set up and only delete the full network when that last workload has expired or has been removed. 
NRs that do nothing don't cost anything anyway.
Checking if a container is connected, also means that somewhere __else__ a container could be connected. And thus if a container is connected, you can't just keep the NR, but you also need to reconfigure the network to reflect the deleted NRs (peers, routes,...)

**Comment by zaibon** [comment]
I don't think you got my point. What you explain involves multiple node. 
I'm only talking about a single node. I'm also not considering a full network. But just a network resource on one node.

> I think it would be better to never expire the network as it is set up and only delete the full network when that last workload has expired or has been removed.

This is not something you can know from a node. And you can perfectly have a same network that have network resource expiring at different time on different node.  The all "network" concept is actually not really relevant anymore.

> Checking if a container is connected, also means that somewhere else a container could be connected. 

I'm also only talking locally on one node, I'm not sahying we need to check if there is other peers that are connected to this network resource. This is not relevant for us. If the network resource expire on nodeA, then all the other peer should loose connection to nodeA. that's it.

> And thus if a container is connected, you can't just keep the NR, but you also need to reconfigure the network to reflect the deleted NRs (peers, routes,...)

This is not up to the node to do something like that by itself. The network is managed by its owner. The node just apply a configuration.

**Comment by delandtj** [comment]
> I don't think you got my point. What you explain involves multiple node.
> I'm only talking about a single node. I'm also not considering a full network. But just a network resource on one node.

the problem is that a network is ONE, with it's NRs, peers, routes. That also means that if NRs are removed form the network, all other NRs need to be aware of that and reconfigure in function of a removed peer and removed routes. Ergo: the TNo needs to be updated somewhere along the line and as such redistributed

> 
> This is not something you can know from a node. And you can perfectly have a same network that have network resource expiring at different time on different node. The all "network" concept is actually not really relevant anymore.
> 
so expiration should be set to infinite when you **create** the network, and when 3bot knows there are no containers any more, it needs to update the TNo with an expiration of **now**
And even then, we can test for existence of containers on that network, and issue a warning that there are still workloads on that network.

> I'm also only talking locally on one node, I'm not sahying we need to check if there is other peers that are connected to this network resource. This is not relevant for us. If the network resource expire on nodeA, then all the other peer should loose connection to nodeA. that's it.
> 
But they **don't** lose the connection, that's the whole point. in the other NRs the routes and endpoints to that deleted NR still exist.

> This is not up to the node to do something like that by itself. The network is managed by its owner. The node just apply a configuration.

So in a nutshell, there is a many to one relationship from containers to a network, and only if no containers, a network can be deleted

**Comment by zaibon** [comment]
> the problem is that a network is ONE, with it's NRs, peers, routes. That also means that if NRs are removed form the network, all other NRs need to be aware of that and reconfigure in function of a removed peer and removed routes. Ergo: the TNo needs to be updated somewhere along the line and as such redistributed

Is it a problem is a network resource has a peer that doesn't respond ? If not, then there is no real "need" to clean up things there.

> so expiration should be set to infinite when you create the network, and when 3bot knows there are no containers any more, it needs to update the TNo with an expiration of now
And even then, we can test for existence of containers on that network, and issue a warning that there are still workloads on that network.

This is not how things works nowadays. Maybe this is something that we should aim for but today there is no management of a TNO at all. User creates a network reservation and that's it. The 3bot has 0 knowledge of network management. 

> But they don't lose the connection, that's the whole point. in the other NRs the routes and endpoints to that deleted NR still exist.

Yes they still exists, but the traffic should be impossible to the network resource that is expired. That's why I propose to close the firewall from any traffic. You should not be able to send traffic to a network resource that is expired.

I think it really comes down to how do you want to manage a network over time for multiple node.
As of today there is no TNO that can change over time and be re-applied. This simply doesn't exist.

A network reservation is like any other reservation a one time description of a primitive that a 0-OS node need to create for a defined period.

If the network change, a new reservation, completely detached from the previous one needs to be created.

In 0-OS network daemon we already support update of a network resource configuration at runtime. This allow to add/remove peer without losing the all connection to other peers. But at the higher management layer, there is no notion of update of existing reservation. Would it be just to extend a expiration time or to modify a network topology.

Now maybe this should make us reflect on how a network should really be managed. And define another interface for it.

**Comment by delandtj** [comment]
but... come to think of it... I wonder how we would have a resource busy when destroying an nr... if there are no processes running in there, everything should vanish at the same time...
show me when that happens, it shouldn't

**Comment by zaibon** [comment]
> if there are no processes running in there, everything should vanish at the same time...

But there are processes running. This is the case when a container network namespace is connected to a network resource and the network resource is destroyed. At that point there is a veth pair connected to the network resource bridge and inside the container network namespace. You can see if in the logs in the first comment.

**Comment by delandtj** [comment]
```
➜  sudo ip netns add test
➜  sudo ip link add test type bridge
➜  sudo ip link add intest type veth  peer name outtest
➜  sudo ip link set intest netns test
➜  sudo ip link set outtest up
➜  sudo ip -n test l set lo up
➜  sudo ip -n test addr add 192.168.123.1/24 dev intest
➜  sudo ip -n test link set intest up
➜  sudo ip -n test r
192.168.123.0/24 dev intest proto kernel scope link src 192.168.123.1
➜  sudo ip link set outtest master test
➜  sudo ip addr add 192.168.123.2/24 dev test
➜  sudo ip l set test up
➜  ping -c1 192.168.123.1
PING 192.168.123.1 (192.168.123.1) 56(84) bytes of data.
64 bytes from 192.168.123.1: icmp_seq=1 ttl=64 time=0.130 ms

--- 192.168.123.1 ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 0.130/0.130/0.130/0.000 ms

➜  sudo ip l
40: test: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP mode DEFAULT group default qlen 1000
    link/ether 8a:1c:08:9d:12:ef brd ff:ff:ff:ff:ff:ff
41: outtest@if42: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master test state UP mode DEFAULT group default qlen 1000
    link/ether ae:da:3e:b6:5f:b8 brd ff:ff:ff:ff:ff:ff link-netns test
~

➜  sudo ip -n test l
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
42: intest@if41: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP mode DEFAULT group default qlen 1000
    link/ether 62:b3:a5:04:88:1a brd ff:ff:ff:ff:ff:ff link-netnsid 0
    
➜  sudo ip netns del test
~ 
➜  ip l
40: test: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue state DOWN mode DEFAULT group default qlen 1000
    link/ether 8a:1c:08:9d:12:ef brd ff:ff:ff:ff:ff:ff
```
so it must be for another reason the the resource is busy, as all vanished, except the bridge (that I didn't delete)

**Comment by zaibon** [comment]
But you have no process running or container using the namespace you have created. I don't think this is a valid example of what happens for real in the node.

What would be the other reason for logs like these  ? :
```
[-] provisiond: 11:55AM INF start decommissioning reservation duration=1h0m0s id=r-0 type=network
[-] capacityd: {"level":"info","time":"2019-10-07T11:55:40Z","message":"send heart-beat to BCDB"}
[  617.221010] device vethbfb0712b left promiscuous mode
[  617.222233] b-5CFQvzwsmVtGM: port 1(vethbfb0712b) entered disabled state
[-] networkd: 11:55AM ERR failed to delete network resource namespace error="failed to unmount NS: at /var/run/netns/n-5CFQvzwsmVtGM: device or resource busy" namespace=n-5CFQvzwsmVtGM
```

**Comment by delandtj** [comment]
.... 5CFQ... is mine, iirc .... I miiiiight have had a `ash` session in that NS, as I can only think of a busy namespace being one with a process running in it...
So, AFAICT, that would be my fault

edit: arf no, my net is `3jyVcwygC5nz2`

**Comment by DylanVerstraete** [comment]
Closing because it's not relevant in 3.0 anymore.
